setup.py

```diff
@@ -1,134 +1,23 @@
-#!/usr/bin/env python
+import subprocess
 
-from wagtail import __version__
-from wagtail.utils.setup import assets, check_bdist_egg, sdist
+from setuptools import setup
+from setuptools.command.sdist import sdist as base_sdist
 
-try:
-    from setuptools import find_packages, setup
-except ImportError:
-    from distutils.core import setup
 
+class sdist(base_sdist):
+    def compile_assets(self):
+        try:
+            subprocess.check_call(["npm", "run", "build"])
+        except (OSError, subprocess.CalledProcessError) as e:
+            print("Error compiling assets: " + str(e))  # noqa: T201
+            raise SystemExit(1)
 
-# Hack to prevent "TypeError: 'NoneType' object is not callable" error
-# in multiprocessing/util.py _exit_function when setup.py exits
-# (see http://www.eby-sarna.com/pipermail/peak/2010-May/003357.html)
-try:
-    import multiprocessing  # noqa: F401
-except ImportError:
-    pass
+    def run(self):
+        self.compile_assets()
+        base_sdist.run(self)
 
 
-install_requires = [
-    "Django>=4.2,<6.0",
-    "django-modelcluster>=6.2.1,<7.0",
-    "django-permissionedforms>=0.1,<1.0",
-    "django-taggit>=5.0,<7",
-    "django-treebeard>=4.5.1,<5.0",
-    "djangorestframework>=3.15.1,<4.0",
-    "django-filter>=23.3",
-    "draftjs_exporter>=2.1.5,<6.0",
-    "Pillow>=9.1.0,<12.0.0",
-    "beautifulsoup4>=4.8,<5",
-    "Willow[heif]>=1.8.0,<2",
-    "requests>=2.11.1,<3.0",
-    "openpyxl>=3.0.10,<4.0",
-    "anyascii>=0.1.5",
-    "telepath>=0.3.1,<1",
-    "laces>=0.1,<0.2",
-    "django-tasks>=0.6.1,<0.7",
-]
-
-# Testing dependencies
-testing_extras = [
-    # Required for running the tests
-    "python-dateutil>=2.7",
-    "Jinja2>=3.0,<3.2",
-    "boto3>=1.28,<2",
-    "freezegun>=0.3.8",
-    "azure-mgmt-cdn>=12.0,<13.0",
-    "azure-mgmt-frontdoor>=1.0,<1.1",
-    "django-pattern-library>=0.7",
-    # For coverage and PEP8 linting
-    "coverage>=3.7.0",
-    "doc8==0.8.1",
-    "ruff==0.9.6",
-    # For enforcing string formatting mechanism in source files
-    "semgrep==1.40.0",
-    # For templates linting
-    "curlylint==0.13.1",
-    # For template indenting
-    "djhtml==3.0.6",
-    # For validating string formats in .po translation files
-    "polib>=1.1,<2.0",
-    # For wagtail.test.utils.wagtail_factories (used for streamfield migration toolkit)
-    "factory-boy>=3.2",
-    # For running tests in parallel
-    "tblib>=2.0,<3.0",
-]
-
-# Documentation dependencies
-documentation_extras = [
-    "pyenchant>=3.1.1,<4",
-    "sphinxcontrib-spelling>=7,<8",
-    "Sphinx>=7.3",
-    "sphinx-autobuild>=0.6.0",
-    "sphinx-wagtail-theme==6.5.0",
-    "myst_parser==2.0.0",
-]
-
 setup(
     name="wagtail",
```

Comment:
Keeping this per https://github.com/wagtail/wagtail/issues/12482#issuecomment-2660343849, but I doubt it does much. I don't see the "Used by" section on the right sidebar of https://github.com/torchbox/django-pattern-library either, so I think it doesn't really work. It should still work in https://github.com/torchbox/django-pattern-library/network/dependents though.

FWIW, Django itself has completely removed setup.py.

In our case, it's still necessary to keep it if we want to use `setuptools` as the build backend and want to run `npm run build` automatically as part of the build process, because we need to override the command via `cmdclass`.